src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs

```diff
@@ -0,0 +1,145 @@
+                catch (Exception e) when (FatalError.ReportWithoutCrash(e))
+                {
+                }
+            }
+
+            if (assemblyLocation == null)
+            {
+                var reference = compilation.GetMetadataReference(symbol.ContainingAssembly);
+                assemblyLocation = (reference as PortableExecutableReference)?.FilePath;
+                if (assemblyLocation == null)
+                {
+                    throw new NotSupportedException(EditorFeaturesResources.Cannot_navigate_to_the_symbol_under_the_caret);
+                }
+            }
+
+            // Decompile
+            document = PerformDecompilation(document, fullName, compilation, assemblyLocation);
+
+            // Convert XML doc comments to regular comments, just like MAS
+            var docCommentFormattingService = document.GetLanguageService<IDocumentationCommentFormattingService>();
+            document = await ConvertDocCommentsToRegularComments(document, docCommentFormattingService, cancellationToken);
+
+            // Parse document
+            var node = await document.GetSyntaxRootAsync();
+
+            // Apply formatting rules
+            document = await Formatter.FormatAsync(
+                  document, SpecializedCollections.SingletonEnumerable(node.FullSpan),
+                  options: null, rules: GetFormattingRules(document), cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return document;
+        }
+
+        private static Document PerformDecompilation(Document document, string fullName, Compilation compilation, string assemblyLocation)
+        {
+            // Load the assembly.
+            var pefile = new PEFile(assemblyLocation, PEStreamOptions.PrefetchEntireImage);
+
+            // Initialize a decompiler with default settings.
+            var decompiler = new CSharpDecompiler(pefile, new AssemblyResolver(compilation), new DecompilerSettings());
+            // Escape invalid identifiers to prevent Roslyn from failing to parse the generated code.
+            // (This happens for example, when there is compiler-generated code that is not yet recognized/transformed by the decompiler.)
+            decompiler.AstTransforms.Add(new EscapeInvalidIdentifiers());
+
+            var fullTypeName = new FullTypeName(fullName);
+
+            var decompilerVersion = FileVersionInfo.GetVersionInfo(typeof(CSharpDecompiler).Assembly.Location);
+
+            // Add header to match output of metadata-only view.
+            // (This also makes debugging easier, because you can see which assembly was decompiled inside VS.)
+            var header = $"#region {FeaturesResources.Assembly} {pefile.FullName}" + Environment.NewLine
+                + $"// {assemblyLocation}" + Environment.NewLine
+                + $"// Decompiled with ICSharpCode.Decompiler {decompilerVersion.FileVersion}" + Environment.NewLine
+                + "#endregion" + Environment.NewLine;
```

Comment:
i'm a little concerned about Environment.NewLine.  But i support the formatter will fix up all of this.  What i want to avoid is any situation where the file is opened and the user is prompted about mismatched newlines.